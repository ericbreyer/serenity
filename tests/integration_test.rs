#![cfg(test)]

use anyhow::Result;
use insta::{assert_debug_snapshot, assert_snapshot};
use test_case::test_case;

#[test_case("trivial.ser", 1; "trivial")]
#[test_case("prime_sieve.ser", 8181807856294299570; "prime_sieve")]
#[test_case("interfaces.ser", 12; "interfaces")]
#[test_case("babbage.ser", 25264; "babbage")]
#[test_case("linkedlist.ser", 55; "linkedlist")]
#[test_case("prime_conspiricy.ser", 7942686168; "prime_conspiricy")]
#[test_case("highly_composites.ser", 355168; "highly_composites")]
pub fn test_file_run(file : &str, ecode: i64) -> Result<()> {

    let file = format!("tests/{}", file);

    let mut out = Vec::new();
    let code = serenity::run_file(&file, &mut out)?;

    assert!(code == ecode, "Expected: {}, Got: {}", ecode, code);    
    
    Ok(())
}

#[test_case("trivial.ser"; "trivial")]
#[test_case("prime_sieve.ser"; "prime_sieve")]
#[test_case("interfaces.ser"; "interfaces")]
#[test_case("babbage.ser"; "babbage")]
#[test_case("linkedlist.ser"; "linkedlist")]
#[test_case("prime_conspiricy.ser"; "prime_conspiricy")]
#[test_case("highly_composites.ser"; "highly_composites")]
pub fn test_file_scan(file : &str) -> Result<()> {
    let file = format!("tests/{}", file);
    let tokens = serenity::scan(&file)?;

    let mut settings = insta::Settings::clone_current();
    settings.set_snapshot_suffix(file);
    let _guard = settings.bind_to_scope();

    assert_debug_snapshot!(tokens);
    Ok(())
}

#[test_case("trivial.ser"; "trivial")]
#[test_case("prime_sieve.ser"; "prime_sieve")]
#[test_case("interfaces.ser"; "interfaces")]
#[test_case("babbage.ser"; "babbage")]
#[test_case("linkedlist.ser"; "linkedlist")]
#[test_case("prime_conspiricy.ser"; "prime_conspiricy")]
#[test_case("highly_composites.ser"; "highly_composites")]
pub fn test_file_parse(file : &str) -> Result<()> {
    let file = format!("tests/{}", file);
    let ast = serenity::parse(&file)?;

    let mut settings = insta::Settings::clone_current();
    settings.set_snapshot_suffix(file);
    let _guard = settings.bind_to_scope();

    assert_snapshot!(ast);
    Ok(())
}

#[test_case("trivial.ser"; "trivial")]
#[test_case("prime_sieve.ser"; "prime_sieve")]
#[test_case("interfaces.ser"; "interfaces")]
#[test_case("babbage.ser"; "babbage")]
#[test_case("linkedlist.ser"; "linkedlist")]
#[test_case("prime_conspiricy.ser"; "prime_conspiricy")]
#[test_case("highly_composites.ser"; "highly_composites")]
pub fn test_file_compile(file : &str) -> Result<()> {
    let file = format!("tests/{}", file);
    let ir = serenity::compile(&file)?;

    let mut settings = insta::Settings::clone_current();
    settings.set_snapshot_suffix(file);
    let _guard = settings.bind_to_scope();

    assert_snapshot!(ir);
    Ok(())
}

