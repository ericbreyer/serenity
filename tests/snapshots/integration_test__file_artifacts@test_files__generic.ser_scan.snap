---
source: tests/integration_test.rs
expression: result
---
<Type         "type"        [1]>
<Less         "<"           [1]>
<Identifier   "T"           [1]>
<Comma        ","           [1]>
<Identifier   "S"           [1]>
<Greater      ">"           [1]>
<Identifier   "s"           [1]>
<Struct       "struct"      [1]>
<LeftBrace    "{"           [1]>
<Identifier   "i"           [2]>
<Colon        ":"           [2]>
<Identifier   "T"           [2]>
<Comma        ","           [2]>
<Identifier   "j"           [3]>
<Colon        ":"           [3]>
<Identifier   "S"           [3]>
<RightBrace   "}"           [4]>
<Semicolon    ";"           [4]>
<Fun          "fn"          [6]>
<Identifier   "main"        [6]>
<LeftParen    "("           [6]>
<RightParen   ")"           [6]>
<RightArrow   "->"          [6]>
<SimpleType   "int"         [6]>
<LeftBrace    "{"           [6]>
<Var          "let"         [7]>
<Identifier   "s_conc"      [7]>
<Colon        ":"           [7]>
<Struct       "struct"      [7]>
<Identifier   "s"           [7]>
<Equal        "="           [7]>
<Struct       "struct"      [7]>
<Identifier   "s"           [7]>
<LeftBrace    "{"           [7]>
<Identifier   "i"           [7]>
<Colon        ":"           [7]>
<Number       "6.2"         [7]>
<Comma        ","           [7]>
<Identifier   "j"           [7]>
<Colon        ":"           [7]>
<Number       "0.0"         [7]>
<RightBrace   "}"           [7]>
<Semicolon    ";"           [7]>
<Var          "let"         [8]>
<Identifier   "s_conc2"     [8]>
<Colon        ":"           [8]>
<Struct       "struct"      [8]>
<Identifier   "s"           [8]>
<Less         "<"           [8]>
<SimpleType   "int"         [8]>
<Comma        ","           [8]>
<SimpleType   "float"       [8]>
<Greater      ">"           [8]>
<Semicolon    ";"           [8]>
<Identifier   "s_conc2"     [9]>
<Dot          "."           [9]>
<Identifier   "i"           [9]>
<Equal        "="           [9]>
<Number       "6"           [9]>
<Semicolon    ";"           [9]>
<Return       "return"      [10]>
<Identifier   "s_conc2"     [10]>
<Dot          "."           [10]>
<Identifier   "i"           [10]>
<Semicolon    ";"           [10]>
<RightBrace   "}"           [11]>
<Eof          ""            [11]>
